flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java

```diff
@@ -651,6 +659,34 @@ private DispatcherGateway getDispatcherGateway() throws LeaderRetrievalException
 		}
 	}
 
+	private void uploadJarFiles(final DispatcherGateway currentDispatcherGateway, final JobGraph job) throws JobExecutionException, InterruptedException {
```

Comment:
Could we pass in the `JobID` and the collection of user jars instead of the `JobGraph`? We would then not set the blob keys in this function but instead return them. That way we would make this method only do the upload instead of multiple things. I think this would better follow the single responsibility principle.